lib/configuration.ex

```diff
@@ -0,0 +1,96 @@
+defmodule Telepoison.Configuration do
+  @moduledoc false
+
+  alias HTTPoison.Request
+
+  @spec setup(infer_fn: (Request.t() -> String.t()), ot_attributes: [{String.t(), String.t()}]) :: :ok
+  def setup(opts \\ []) do
+    Agent.start_link(fn -> set_defaults(opts) end, name: __MODULE__)
+
+    :ok
+  end
+
+  defp set_defaults(opts) do
+    infer_fn =
+      case Keyword.get(opts, :infer_route) do
+        nil ->
+          &Telepoison.URI.infer_route_from_request/1
+
+        infer_fn when is_function(infer_fn, 1) ->
+          infer_fn
+      end
+
+    ot_attributes =
+      case Keyword.get(opts, :ot_attributes) do
+        ot_attributes when is_list(ot_attributes) ->
+          Enum.map(ot_attributes, fn
+            {key, value} when is_binary(key) and is_binary(value) ->
+              {key, value}
+
+            _ ->
+              nil
+          end)
+
+        _ ->
+          []
+      end
+
+    {infer_fn, ot_attributes}
+  end
+
+  @doc """
+  Get a configuration value or raise `ArgumentError`
+  """
+  @spec get!(:infer_fn | :ot_attributes) :: any
+  def get!(key) do
+    case get(key) do
+      {:ok, value} -> value
+      {:error, error} -> raise ArgumentError, error
+    end
+  end
+
+  @doc """
+  Get a configuration value if present and `setup` has been called or return an error
+  """
+  @spec get(:infer_fn | :ot_attributes) :: {:ok, any()} | {:error, String.t()}
+  def get(key)
+
+  def get(:infer_fn) do
+    if agent_started?() do
+      Agent.get(
+        __MODULE__,
+        fn
+          {infer_fn, _} when is_function(infer_fn, 1) ->
```

Comment:
shouldn't you do these kind of configuration validations during setup (fail early)?